reinitialize_ function for refactoring Tensor as member variable

jerryzh168 opened the pull request:
Summary:
We want to refactor
```
class A {

void func() {
  x_.Resize(dims);
  auto* data = x_.mutable_data<T>();
}

Tensor x(CPU);
};
```

to
```
class A {
void func() {
  reinitialize_(&x_, dims, at::dtype<T>().device(CPU));
  auto* data = x_.mutable_data<T>();
}

Tensor x_; // Undefined Tensor
};
```

This diff adds the reinitialize_ function.

Differential Revision: D10861298

ezyang commented on caffe2/core/tensor.cc:
```diff
@@ -124,6 +124,34 @@ Tensor empty(at::IntList dims, at::TensorOptions options) {
+    Tensor* tensor,
+    at::IntList dims,
+    at::TensorOptions options) {
+  CAFFE_ENFORCE(options.device_opt() != c10::nullopt);
```
OOC, why not use the defaulting? (`options.device()`)

jerryzh168 replied:
because of the specific use case we are expecting(for actual initialization), I think it might be better to enforce users to provide device type? I don't have strong opinion on this tho, we may say when we use it in c2 code base we'll always provide it as well.

dzhulgakov replied:
given that we want to use in C2 ops for now - it's better to make it explicit indeed

ezyang commented on caffe2/core/tensor.cc:
```diff
@@ -124,6 +124,34 @@ Tensor empty(at::IntList dims, at::TensorOptions options) {
+      if (tensor->meta() == type_meta) {
+        tensor->raw_mutable_data();
+      } else {
+        // create a new Tensor when the data_type doesn't match
```
These dtype changes are going to keep giving me nightmares.

dzhulgakov replied:
we could start throwing a WARNING here if the type was initialized previously and see how often it gets triggered

jerryzh168 replied:
For current codemod, the pattern we are matching is Tensor in memeber variable and we'll replace the first Resize call with initialize, I don't think it will get triggered. I think the question is whether we want to use this API for other purposes as well, or just for the single purpose of reinitializing a member variable of type Tensor.

ezyang commented:
I'm not sure this function should handle dtype-change. In the two(!!) cases of caching I've seen, they only check for size change, not dtype change. This implies that the dtype never changes. I'm much happier if we don't have to deal with a dtype change in this function.

Also, I'd strongly advise you to write a description of this function that makes sense, independently of the refactor you're trying to do. If it seems hard or convoluted to describe what it does, that implies that maybe we should do more work figuring out a clear API semantics.

jerryzh168 commented:
Yeah, I agree that we should define a clear semantics for this function first. My initial thought was just to use this as a delayed initialization, but we could also support broader uses cases like use this for Resize when we provide a different dims etc.

dzhulgakov commented on caffe2/core/tensor.cc:
```diff
@@ -124,6 +124,34 @@ Tensor empty(at::IntList dims, at::TensorOptions options) {
   return tensor;
 }
 
+void reinitialize_(
```
probably can benefit from a better name? ReinitializeBuffer or something?

jerryzh168 replied:
Sure

jerryzh168 replied:
I'll change it to ReinitializeTensor

jerryzh168 committed:
ReinitializeTensor function for refactoring Tensor as member variable (#13147)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/13147

We want to refactor
```
class A {

void func() {
  x_.Resize(dims);
  auto* data = x_.mutable_data<T>();
}

Tensor x(CPU);
};
```

to
```
class A {
void func() {
  ReinitializeTensor(&x_, dims, at::dtype<T>().device(CPU));
  auto* data = x_.mutable_data<T>();
}

Tensor x_; // Undefined Tensor
};
```

This diff adds the ReinitializeTensor function.

Reviewed By: dzhulgakov

Differential Revision: D10861298

fbshipit-source-id: 0cd09169abefcb0b819bfc897f92f5d5c125388e
